staging/src/k8s.io/apiserver/pkg/cel/library/network_addrs.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package library
+
+import (
+	"net/netip"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+// NetworkAddr provides a CEL function library extension of network address parsing functions.
+//
+// networkAddr
+//
+// Converts a string to a network address or results in an error if the string is not a valid network address.
+// The network address must be an IPv4 or IPv6 address with a CIDR mask.
+//
+//	networkAddr(<string>) <NetworkAddr>
+//
+// Examples:
+//
+//	networkAddr('192.168.0.0/16') // returns an IPv4 address with a CIDR mask
+//	networkAddr('::1/128') // returns an IPv6 address with a CIDR mask
+//	networkAddr('192.168.0.0/33') // error
+//	networkAddr('::1/129') // error
+//
+// isNetworkAddr
+//
+// Returns true if a string is a valid network address. The network address must be an IPv4 or IPv6 address with a CIDR mask.
+//
+//	isNetworkAddr(<string>) <bool>
+//
+// Examples:
+//
+//	isNetworkAddr('192.168.0.0/16') // returns true
+//	isNetworkAddr('::1/128') // returns true
+//	isNetworkAddr('192.168.0.0/33') // returns false
+//	isNetworkAddr('::1/129') // returns false
+//
+// containsIPAddr
+//
+// - containsIPAddr: Returns true if a network address contains an IP address. Should be passed an IP address parsed with ipAddr.
```

Comment:
just "containsIP" ?

FWIW, (a) we might want `containsCIDR` too, and (b) it's possible that inverting them (having `IPAddr.isInCIDR` like you'd thought originally, and `NetworkAddr.isInCIDR`) might be more useful (or maybe have it available both ways?). The one place in core k8s I can think of right now where we might use this is in [NetworkPolicy `ipBlock` validation](https://github.com/kubernetes/kubernetes/blob/v1.28.4/pkg/apis/networking/validation/validation.go#L239), where given:

```
- ipBlock
    cidr: 192.168.1.0/24
    except:
      - 192.168.1.1/32
```

you want to check that each `except` value `isInCIDR(cidr)`, which seems easier to check in that direction than in the reverse?

Reply:
> just "containsIP" ?

I like `containsIPAddr` because it's then obvious that you should be using the `ipAddr` function to parse the IP before you pass it into the function. If we decide to change things around this may go away though.

---

I think it doesn't really matter, for both of these, you need to parse the object and validate it. So you'd either have (I'm renaming to CIDR so examples using new names)
```
isCIDR(self.cidr) && self.except.all(x, isIPAddr(x) && cidr(self.cidr).containsIPAddr(x))
```
or
```
isCIDR(self.cidr) && self.except.all(x, isIPAddr(x) && ipAddr(x).isInCIDR(cidr(self.cidr)))
```
Unless there's an easier way to write one of those?

The benefit I guess of doing `ipAddr.isInCIDR` would be then you have `isInCIDR` on both, rather than having `containsCIDR` and `containsIPAddr` both on the `cidr` object.

Will look to other examples to see if there's a preferential style 

--- 

I think we want to have only 1 way to represent an expression, so having both I'm a downvote on, but which way we do this (`cidr.containsIPAddr`/`cidr.containsCIDR`) vs (`ipAddr.isInCIDR`/`cidr.isInCIDR`) I'm not strongly opinionated on

Reply:
ah, I was thinking only of simpler cel, where the rule would be on the `except` field rather than being able to express it at a higher level